src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultTagHelperCompletionService.cs

```diff
@@ -0,0 +1,153 @@
+using System.Linq;
+using Microsoft.AspNetCore.Razor.Language;
+
+namespace Microsoft.VisualStudio.LanguageServices.Razor
+{
+    [Export(typeof(TagHelperCompletionService))]
+    internal class DefaultTagHelperCompletionService : TagHelperCompletionService
+    {
+        [Import(typeof(TagHelperFactsService))]
+        public TagHelperFactsService TagHelperFactsService { get; set; }
+
+        public override ElementCompletionResult GetElementCompletions(ElementCompletionContext completionContext)
+        {
+            if (completionContext == null)
+            {
+                throw new ArgumentNullException(nameof(completionContext));
+            }
+
+            var elementCompletions = new Dictionary<string, IEnumerable<TagHelperDescriptor>>(StringComparer.OrdinalIgnoreCase);
+
+            AddAllowedChildrenCompletions(completionContext, elementCompletions);
+
+            if (elementCompletions.Count > 0)
+            {
+                // If there are allowed children, those are the only possible completions.
+                var emptyResult = ElementCompletionResult.Create(elementCompletions);
+                return emptyResult;
+            }
+
+            elementCompletions = completionContext.ExistingCompletions.ToDictionary(
+                completion => completion,
+                _ => Enumerable.Empty<TagHelperDescriptor>(),
+                StringComparer.OrdinalIgnoreCase);
+
+            var possibleChildDescriptors = TagHelperFactsService.GetTagHelpersGivenParent(completionContext.DocumentContext, completionContext.ContainingTagName);
+            foreach (var possibleDescriptor in possibleChildDescriptors)
+            {
+                var addRuleCompletions = false;
+                var outputHint = possibleDescriptor.TagOutputHint;
+
+                // Filter out catch-all rules because TagHelpers that target attributes only would light up every child tag otherwise. Force those TagHelpers
+                // to have additional requirements before showing them in the element completion list.
+                var nonCatchAllRules = possibleDescriptor.TagMatchingRules.Where(rule => rule.TagName != TagHelperMatchingConventions.ElementCatchAllName);
+                foreach (var rule in nonCatchAllRules)
+                {
+                    if (elementCompletions.ContainsKey(rule.TagName))
+                    {
+                        addRuleCompletions = true;
+                    }
+                    else if (outputHint != null && elementCompletions.ContainsKey(outputHint))
+                    {
+                        // If the possible descriptors final output tag already exists in our list of completions, we should add every representation
+                        // of that descriptor to the possible element completions.
+                        addRuleCompletions = true;
+                    }
+                    else if (!completionContext.InHTMLSchema(rule.TagName))
```

Comment:
Do we really need to know about the HTML schema? Would it be better if we didn't do this filtering and they applied it after?

I don't have a big problem with this I just wanted to make sure we're not taking too big of a dependency on WTE.